daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -223,23 +223,21 @@ private[lf] case class PartialTransaction(
   def finish(
       outputTransactionVersions: VersionRange[TransactionVersion],
       packageLanguageVersion: Ref.PackageId => LanguageVersion,
-  ): Either[PartialTransaction, SubmittedTransaction] = {
-
-    Either.cond(
-      context.exeContext.isEmpty && aborted.isEmpty,
-      SubmittedTransaction(
-        TransactionVersions
-          .asVersionedTransaction(
-            outputTransactionVersions,
-            packageLanguageVersion,
-            context.children.toImmArray,
-            nodes
-          )
-          .fold(SError.crash, identity)
-      ),
-      this
-    )
-  }
+  ): Either[String, Either[PartialTransaction, SubmittedTransaction]] =
```

Comment:
I find nested eithers pretty confusing. Maybe time for a custom ADT for the error types.

Reply:
I will.